File: Python/RenameAAAShpFromAttributeTable.py
# -*- coding: utf-8 -*-
# --------------------------------------------------------------------
# Name: renameAAAShpFromAttributeTable.py
# Renames AAA-shapefiles according to information in the attribute table in order to achieve standard naming
# --------------------------------------------------------------------

import arcpy
import os
import sys
import shutil
import fnmatch
import csv

provDat = arcpy.GetParameterAsText(0)
overwriteExisting = arcpy.GetParameterAsText(1)
pyScript = sys.argv[0]


# Get table path
thisScriptPath = os.path.dirname(pyScript)
rScriptPath = os.path.join(os.path.abspath(os.path.join(thisScriptPath, os.pardir)), "R")
renF = []
with open(os.path.join(rScriptPath, "renameAAAData.csv"), 'r') as csvfile:
    csvObj = csv.reader(csvfile, delimiter=';')
    for row in csvObj:
        renF.append(row)


# get AAA-data folders
aaaFolders = []
if os.path.isdir(provDat + "\\ALKIS"):
    aaaFolders.append(provDat + "\\ALKIS")
if os.path.isdir(provDat + "\\ATKIS"):
    aaaFolders.append(provDat + "\\ATKIS")

for aaaFolder in aaaFolders:
    
    # get all files in directory and check if names abide by the standard naming
    allShp = []
    isStandardNaming = False
    for root, dirnames, filenames in os.walk(aaaFolder):
        for filename in fnmatch.filter(filenames, "*.shp"):
            allShp.append(os.path.join(root, filename))
            if (fnmatch.fnmatch(os.path.join(root, filename), "*\\AX_*")):
                isStandardNaming = True
        
    if isStandardNaming == False:
    
        # create a safety copy to work on
        newFolder = aaaFolder + "_rename"
        
        if ((os.path.isdir(newFolder) == False) or (overwriteExisting == "true")):
            
            if os.path.isdir(newFolder):
                shutil.rmtree(newFolder)
            
            # create new folder    
            os.makedirs(newFolder)
            
            
            # iterate over each shapefile
            for s in range(0,len(allShp)):
                thisShp = allShp[s]
                
                arcpy.AddMessage("Renaming AAA-file " + thisShp + " (" + str(s) + "/" + str(len(allShp)) + ")")
                
                # get field names
                fields = arcpy.ListFields(thisShp)
                
                # get field name which fit
                for thisField in fields:
                    for thisName in renF:
                        if thisField.name == thisName[0]:
                            useField = thisField
                            useName = thisName[0]
                            fieldType = useField.type
                
                
                            # get unique field values            
                            values = [row[0] for row in arcpy.da.SearchCursor(thisShp,(useName))]
                            uniqueValues = list(set(values))
                            
                            # extract features and write to new shapefile
                            for uf in range(0, len(uniqueValues)):
                                thisFeat = uniqueValues[uf]
                                
                                # set destination file and intermediate file paths
                                toPath = newFolder + "\\" + thisFeat + ".shp"
                                interm = newFolder + "\\intermediate.shp"
                                
                                if fieldType == "String":
                                    selectStatement = '"' + useName + '" = ' + "'" + str(thisFeat) + "'"
                                else:
                                    selectStatement = '"' + useName + '" = ' + str(thisFeat)
                                    
                                if os.path.isfile(toPath):
                                    if (os.path.isfile(interm)):
                                        arcpy.Delete_management(interm)
                                    arcpy.Select_analysis(thisShp, interm, selectStatement)
                                    arcpy.Append_management(interm, toPath, "NO_TEST")
                                else:
                                    arcpy.Select_analysis(thisShp, toPath, selectStatement)
                                    

                                # delete intermediate files
                                if (os.path.isfile(interm)):
                                    arcpy.Delete_management(interm)
            
 
# send parameter to ArcGIS
arcpy.SetParameterAsText(2, provDat)

File: R/RenameAAAShpFromAttributeTable.r
# Data management
# Renames AAA-shapefiles according to information in the attribute table in order to achieve standard naming
# Christian Mueller, christian1.mueller@hs-bochum.de, Bochum University of Applied Sciences
# requiered packages: foreign

# user input
# aaaFolder <- 'C:\\HochschuleBochum\\Daten\\Herdecke\\ATKIS_rename'
# wd <- "C:\\HochschuleBochum\\CodesScripts\\R"

# input from python
args <- commandArgs()
aaaFolder <- args[5]
wd <- args[6]


# rearrange data paths
aaaFolder <- gsub(pattern = "\\", replacement = "/", x = aaaFolder, fixed = T)
wd <- gsub(pattern = "\\", replacement = "/", x = wd, fixed = T)

# set working directory
setwd(wd)


# load libraries
print("Loading R-packages...")
for (i in 1:2){
  if (!require("foreign")) install.packages("foreign")
}


# read field name information
fieldNames <- read.table("renameAAAData.csv", sep = ";", dec = ".", header = T, stringsAsFactors = F)[,1]


# get all shapefiles in folder
allShp <- list.files(aaaFolder, pattern = "\\.shp$", full.names = T, recursive = T)

for (s in 1:length(allShp)){
  
  thisShp <- allShp[s]
  
  # AAA-Shapefile name should have an '_' as the third symbol (e.g. AX_Strasse.shp)
  if (((substr(basename(thisShp), 3, 3)) == "_") == F){
    
    dbfPath <- paste(strsplit(thisShp, ".", fixed = T)[[1]][1], "dbf", sep = ".")
    dat <- read.dbf(dbfPath, as.is = T)
    
    # match field names with names known to hold information about the standard AAA-naming
    pos <- which(names(dat) %in% fieldNames)[1]
    
    if (is.na(pos) == F){
    
      # get standard name
      newName <- dat[1,pos]
      if (substr(newName, (nchar(newName) - 3), nchar(newName)) != ".shp"){
        newName <- paste(newName, "shp", sep = ".")
      }
      
      # get all files for this shapefile
      allFiles <- list.files(aaaFolder, pattern = strsplit(basename(thisShp), ".", fixed = T)[[1]][1], full.names = T, recursive = T)
      
      # rename all associated files
      for (f in 1:length(allFiles)){
        
        newPath = paste(dirname(allFiles[f]), "/", strsplit(newName, ".", fixed = T)[[1]][1], ".", strsplit(basename(allFiles[f]), ".", fixed = T)[[1]][2], sep = "")
        
        file.rename(from = allFiles[f], to = newPath)
        
      }
      
    
    }
    
  }
  
}
  
print("...completed execution of R-script")
